image_view/src/image_saver.cpp

```diff
@@ -0,0 +1,284 @@
+
+    // Useful when CameraInfo is being published
+    cam_sub_ = image_transport::create_camera_subscription(
+      this, topic, std::bind(
+        &ImageSaverNode::callbackWithCameraInfo, this, std::placeholders::_1, std::placeholders::_2),
+      "raw");
+
+    // Useful when CameraInfo is not being published
+    image_sub_ = image_transport::create_subscription(
+      this, topic, std::bind(
+        &ImageSaverNode::callbackWithoutCameraInfo, this, std::placeholders::_1),
+      "raw");
+
+    std::string format_string;
+    format_string = this->declare_parameter("filename_format", std::string("left%04i.%s"));
+    encoding = this->declare_parameter("encoding", std::string("bgr8"));
+    save_all_image = this->declare_parameter("save_all_image", true);
+    request_start_end = this->declare_parameter("request_start_end", false);
+    g_format.parse(format_string);
+
+    if (request_start_end && !save_all_image) {
+      RCLCPP_WARN(
+        this->get_logger(), "'request_start_end' is true, so overwriting 'save_all_image' as true");
+    }
+
+    save_srv_ = this->create_service<std_srvs::srv::Empty>(
+      "save", std::bind(&ImageSaverNode::service, this, _1, _2, _3));
+
+    // FIXME(unkown): This does not make services appear
+    // if (request_start_end) {
+    start_srv_ = this->create_service<std_srvs::srv::Trigger>(
+        "start", std::bind(&ImageSaverNode::callbackStartSave, this, _1, _2, _3));
+    end_srv_ = this->create_service<std_srvs::srv::Trigger>(
+        "end", std::bind(&ImageSaverNode::callbackEndSave, this, _1, _2, _3));
+    // }
+  }
+
+  bool service(
+    const std::shared_ptr<rmw_request_id_t> request_header,
+    const std::shared_ptr<std_srvs::srv::Empty::Request> request,
+    std::shared_ptr<std_srvs::srv::Empty::Response> response)
+  {
+    (void)request_header;
+    (void)request;
+    (void)response;
+    save_image_service = true;
+    return true;
+  }
+
+  bool callbackStartSave(
+    const std::shared_ptr<rmw_request_id_t> request_header,
+    const std::shared_ptr<std_srvs::srv::Trigger::Request> request,
+    std::shared_ptr<std_srvs::srv::Trigger::Response> response)
+  {
+    (void)request_header;
+    (void)request;
+    RCLCPP_INFO(this->get_logger(), "Received start saving request");
+    start_time_ = this->now();
+    end_time_ = rclcpp::Time(0);
+
+    response->success = true;
+    return true;
+  }
+
+  bool callbackEndSave(
+    const std::shared_ptr<rmw_request_id_t> request_header,
+    const std::shared_ptr<std_srvs::srv::Trigger::Request> request,
+    std::shared_ptr<std_srvs::srv::Trigger::Response> response)
+  {
+    (void)request_header;
+    (void)request;
+    RCLCPP_INFO(this->get_logger(), "Received end saving request");
+    end_time_ = this->now();
+
+    response->success = true;
+    return true;
+  }
+
+  void callbackWithoutCameraInfo(const sensor_msgs::msg::Image::ConstSharedPtr& image_msg)
```

Comment:
`&` I'm surprised that links, I thought it was only ConstSharedPtr was the profile (but if keeping you're missing a space).

Reply:
Fixed.